README.md
# Logstash Plugin

[![Build Status](https://travis-ci.org/Transrian/logstash-filter-ldap.svg?branch=master)](https://travis-ci.org/Transrian/logstash-filter-ldap)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

**logstash-filter-ldap** filter will add fields queried from a ldap server to the event.
The fields will be stored in a variable called **target**, that you can modify in the configuration file.

If an error occurs during the process tha **tags** array of the event is updated with either:
- **LDAP_ERROR** tag: Problem while connecting to the server: bad *host, port, username, password, or search_dn* -> Check the error message and your configuration.
- **LDAP_NOT_FOUND** tag: Object wasn't found.

If error logging is enabled a field called **error** will also be added to the event.
It will contain more details about the problem.

## Example

### Basic sample

#### Input event

```ruby
{
    "@timestamp" => 2018-02-25T10:04:22.338Z,
    "@version" => "1",
    "myUid" => "u501565"
}
```

#### Logstash filter

```ruby
filter {
  ldap {
    identifier_value => "%{myUid}"
    host => "my_ldap_server.com"
    ldap_port => "389"
    username => "<connect_username>"
    password => "<connect_password>"
    search_dn => "<user_search_pattern>"
  }
}
```

#### Output event

```ruby
{
    "@timestamp" => 2018-02-25T10:04:22.338Z,
    "@version" => "1",
    "myUid" => "u501565",
    "ldap" => {
        "givenName" => "VALENTIN",
        "sn" => "BOURDIER"
    }
}
```

## Parameters availables

Here is a list of all parameters, with their default value, if any, and their description.

|    Option name        | Type    | Required | Default value  | Description                                                                                                   | Example                            |
|:---------------------:|---------|----------|----------------|---------------------------------------------------------------------------------------------------------------|------------------------------------|
| identifier_value      | string  | yes      | n/a            | Identifier of the value to search. If identifier type is uid, then the value should be the uid to search for. | "123456"                           |
| identifier_key        | string  | no       | "uid"          | Type of the identifier to search                                                                              | "uid"                              |
| identifier_type       | string  | no       | "posixAccount" | Object class of the object to search                                                                          | "person"                           |
| search_dn             | string  | yes      | n/a            | Domain name in which search inside the ldap database (usually your userdn or groupdn)                         | "dc=example,dc=org"                |
| attributes            | array   | no       | []             | List of attributes to get. If not set, all attributes available will be get                                   | ['givenName', 'sn']                |
| target                | string  | no       | "ldap"         | Name of the variable you want the result being stocked in                                                     | "myCustomVariableName"             |
| host                  | string  | yes      | n/a            | LDAP server host adress                                                                                       | "ldapserveur.com"                  |
| ldap_port             | number  | no       | 389            | LDAP server port for non-ssl connection                                                                       | 400                                |
| ldaps_port            | number  | no       | 636            | LDAP server port for ssl connection                                                                           | 401                                |
| use_ssl               | boolean | no       | false          | Enable or not ssl connection for LDAP  server. Set-up the good ldap(s)_port depending on that                 | true                               |
| enable_error_logging  | boolean | no       | false          | When there is a problem with the connection with the LDAP database, write reason in the event                 | true                               |
| no_tag_on_failure     | boolean | no       | false          | No tags are added when an error (wrong credentials, bad server, ..) occur                                     | true                               |
| username              | string  | no       | n/a            | Username to use for search in the database                                                                    | "cn=SearchUser,ou=person,o=domain" |
| password              | string  | no       | n/a            | Password of the account linked to previous username                                                           | "123456"                           |
| use_cache             | boolean | no       | true           | Choose to enable or not use of buffer                                                                         | false                              |
| cache_type            | string  | no       | "memory"       | Type of buffer to use. Currently, only one is available, "memory" buffer                                      | "memory"                           |
| cache_memory_duration | number  | no       | 300            | Cache duration (in s) before refreshing values of it                                                          | 3600                               |
| cache_memory_size     | number  | no       | 20000          | Number of object max that the buffer can contains                                                             | 100                                |
| disk_cache_filepath   | string  | no       | nil            | Where the cache will periodically be dumped                                                                     | "/tmp/my-memory-backup"            |
| disk_cache_schedule   | string  | no       | 10m            | Cron period of when the dump of the cache should occured. See [here](https://github.com/floraison/fugit) for the syntax.                           | "10m", "1h", "every day at five", "3h10m"          |

## Buffer

Like all filters, this filter treat only 1 event at a time.
This can lead to some slowing down of the pipeline speed due to the network round-trip time, and high network I/O.

A buffer can be set to mitigate this.

Currently, there is only one basic **"memory"** buffer.

You can enable / disable use of buffer with the option **use_cache**.

### Memory Buffer

This buffer **store** data fetched from the LDAP server **in RAM**, and can be configured with two parameters:
- *cache_memory_duration*: duration (in s) before a cache entry is refreshed if hit.
- *cache_memory_size*: number of tuple (identifier, attributes) that the buffer can contains.

Older cache values than your TTL will be removed from cache.

## Persistant cache buffer

For the only buffer for now, you will be able to save it to disk periodically.

Some specificities :
  - for *the memory cache*, TTL will be reset

Two parameters are required: 
  - *disk_cache_filepath*: path on disk of this backup
  - *disk_cache_schedule*: schedule (every X time unit) of this backup. Please check [here](https://github.com/floraison/fugit) for the syntax of this parameter. 

## Development

If you want to help developing this plugin, you can use the [Vagrantfile](Vagrantfile) to setup your environment.
Requirements :
- [Vagrant](https://www.vagrantup.com/)
- [VirtualBox](https://www.virtualbox.org/)

Here are the steps:

``` bash
# Create the VM, and provision it
vagrant up

# Connect with SSH to the VM
vagrant ssh

# Go inside the project directory
$ cd /vagrant

# Download ruby dependencies
$ bundle install

# Execute tests
$ bundle exec rspec

# Build the Gemfile
$ gem build logstash-filter-ldap.gemspec

# Generate the offline-archive
# You need to install the Logstash yourself
$LOGSTASH_HOME/bin/logstash-plugin install --no-verify logstash-filter-ldap-X.X.X.gem
$LOGSTASH_HOME/bin/logstash-plugin prepare-offline-pack logstash-filter-ldap
```

### Proxy configuration

If you use a proxy, such as a corporate proxy, here are the steps:
- Export your proxy settings: you need to setup these environment variables (even on Windows !):
  - **http_proxy** (eg. http_proxy="http://192.168.0.2:3128/")
  - **https_proxy** (eg. https_proxy="http://192.168.0.2:3128/")
  - **no_proxy** (eg. no_proxy="localhost,127.0.0.1,.example.com")
- Install the Vagrant [proxyconf](https://github.com/tmatilai/vagrant-proxyconf) plugin:
  > vagrant plugin install vagrant-proxyconf

And that's all, all should work!

## Thanks for

This plugin was strongly inspired by the [logstash_filter_LDAPresolve](https://github.com/EricDeveaud/logstash_filter_LDAPresolve), made by [EricDeveaud](https://github.com/EricDeveaud).

## TODO

Instead creating one connection for each event, create only one into the instance class, with a retry number setup into the configuration file.

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 0.1.0
  - Plugin created with the logstash plugin generator
## 0.2.0
  - Rename of some config fields
## 0.2.1
  - Changed library for ldap queries
  - Fixed bugs concerning LDAPs connections
## 0.2.2
  - Added a no_tag_on_failure option
## 0.2.3
  - avoid hash computation if the cache is not required
  - we now use [LRU Cache](https://github.com/SamSaffron/lru_redux) as default memory caching algorithm
  - memory cache is enabled by default
## 0.2.4
  - Fix non atomic cache operation when getting a cache result
  - Updated net-ldap library version
  - We can periodicly save buffer to disk

spec/filters/ldap_spec.rb
# encoding: utf-8

require_relative '../spec_helper'
require "logstash/filters/ldap"

# We disable warning for warning of others dependencies
$VERBOSE = nil

describe LogStash::Filters::Ldap do

  # You need to set-up all those environement variables to
  # test this plugin using "bundle exec rspect"
  before(:each) do
    @ldap_host=ENV["ldap_host"]
    @ldap_port=ENV["ldap_port"]
    @ldaps_port=ENV["ldaps_port"]
    @ldap_username=ENV["ldap_username"]
    @ldap_password=ENV["ldap_password"]
    @ldap_search_dn=ENV["ldap_search_dn"]
  end


  describe "check simple search" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("dn")
      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end

    sample("test" => "test2" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end

  describe "check simple search with persistant cache" do

    cache_path = "/tmp/test_cache"
    File.delete(cache_path) if File.exist?(cache_path)

    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
          use_cache => true
          disk_cache_filepath => "#{@cache_path}"
          disk_cache_schedule => "1"
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
    end

    sample("test" => "test" ) do
      sleep(10)
      expect(Pathname(cache_path).exist?).to eq(true)
    end

  end

  describe "check simple search without attributes" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('homedirectory')
      expect(subject.get('ldap')).to include('gecos')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["gecos"]).to eq("VALENTIN BOURDIER")
      expect(subject.get("ldap")["homedirectory"]).to eq("/users/login/u501565")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end


  describe "check simple search with ssl" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          use_ssl => true
          ldaps_port => "#{@ldaps_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end

    sample("test" => "test2" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end

  describe "check simple search with cache" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
          use_cache => "true"
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end

    sample("test" => "test2" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end


  describe "check simple search with custom object type" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          identifier_type => "person"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end

  describe "check with false ssl settings" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          use_ssl => true
          host => "#{@ldap_host}"
          ldaps_port => "1234"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          enable_error_logging => true
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include("error")
      expect(subject).to include('tags')

      expect(subject.get('ldap')).not_to include('givenname')
      expect(subject.get('ldap')).not_to include('sn')

      expect(subject.get("tags")).to eq(["LDAP_ERROR"])
      expect(subject.get("ldap")["error"]).to match("Bad file descriptor - No message available")
    end
  end


  describe "check simple search with custom identifier" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_key => "homeDirectory"
          identifier_value => "/users/login/u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('givenname')
      expect(subject.get('ldap')).to include('sn')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["givenname"]).to eq("VALENTIN")
      expect(subject.get("ldap")["sn"]).to eq("BOURDIER")
    end
  end


  describe "check simple search with customs attributs" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ["cn", "uidnumber", "gidnumber"]
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include('cn')
      expect(subject.get('ldap')).to include('uidnumber')
      expect(subject.get('ldap')).to include('gidnumber')

      expect(subject.get('ldap')).not_to include('givenname')
      expect(subject.get('ldap')).not_to include('sn')
      expect(subject.get('ldap')).not_to include("error")
      expect(subject.get('ldap')).not_to include('tags')

      expect(subject.get('ldap')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("ldap")["cn"]).to eq("VALENTIN BOURDIER - U501565")
      expect(subject.get("ldap")["uidnumber"]).to eq("479615")
      expect(subject.get("ldap")["gidnumber"]).to eq("9043")
    end
  end


  describe "check bad ldap host" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "babdsfafds.org"
          ldap_port => "#{@ldap_port}"
          username => "test"
          password => "test"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
          enable_error_logging => true
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include("error")
      expect(subject).to include('tags')

      expect(subject.get('ldap')).not_to include('givenname')
      expect(subject.get('ldap')).not_to include('sn')

      expect(subject.get("tags")).to eq(["LDAP_ERROR"])
      expect(subject.get("ldap")["error"]).to eq("getaddrinfo: name or service not known")
    end
  end


  describe "check bad ldap host without error logging" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "babdsfafds.org"
          ldap_port => "#{@ldap_port}"
          username => "test"
          password => "test"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('tags')

      expect(subject).not_to include('ldap')

      expect(subject.get("tags")).to eq(["LDAP_ERROR"])
    end
  end


  describe "test bad search_dn" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          attributes => ['givenName', 'sn']
          search_dn => "test"
          enable_error_logging => true
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include("error")
      expect(subject).to include('tags')

      expect(subject.get('ldap')).not_to include('givenname')
      expect(subject.get('ldap')).not_to include('sn')

      expect(subject.get("tags")).to eq(["LDAP_ERROR"])
      expect(subject.get("ldap")["error"]).to eq("invalid DN")
    end
  end


  describe "test bad user/password couple" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "test"
          password => "test"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
          enable_error_logging => true
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('ldap')

      expect(subject.get('ldap')).to include("error")
      expect(subject).to include('tags')

      expect(subject.get('ldap')).not_to include('givenname')
      expect(subject.get('ldap')).not_to include('sn')

      expect(subject.get("tags")).to eq(["LDAP_ERROR"])
      expect(subject.get("ldap")["error"]).to eq("invalid DN")
    end
  end

  describe "test bad user/password couple & no_tag" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "test"
          password => "test"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
          no_tag_on_failure => true
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).not_to include('ldap')
      expect(subject).not_to include('tags')
    end
  end

  describe "check bad identifier user" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "abcdefg"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('tags')

      expect(subject).not_to include('ldap')

      expect(subject.get("tags")).to eq(["LDAP_NOT_FOUND"])
    end
  end


  describe "check simple search with custom target" do
    let(:config) do <<-CONFIG
      filter {
        ldap {
          identifier_value => "u501565"
          target => "myTarget"
          host => "#{@ldap_host}"
          ldap_port => "#{@ldap_port}"
          username => "#{@ldap_username}"
          password => "#{@ldap_password}"
          search_dn => "#{@ldap_search_dn}"
          attributes => ['givenName', 'sn']
        }
      }
      CONFIG
    end

    sample("test" => "test" ) do
      expect(subject).to include('myTarget')

      expect(subject.get('myTarget')).to include('givenname')
      expect(subject.get('myTarget')).to include('sn')

      expect(subject.get('myTarget')).not_to include("error")
      expect(subject).not_to include('tags')

      expect(subject.get("myTarget")["givenname"]).to eq("VALENTIN")
      expect(subject.get("myTarget")["sn"]).to eq("BOURDIER")
    end
  end


end


spec/filters/cache_saver_spec.rb
# encoding: utf-8
require_relative '../spec_helper'
require "logstash/filters/ldap"

describe "Test cache saver" do

  before(:each) do
    @default_cache_filepath= "/tmp/test_cache"
    @default_data = {
      :a => "fds",
      :b => "aaz"
    }
  end


  it "simple cache -> uncache process" do
    cache = CacheSaver.new(@default_cache_filepath)

    succeed, error = cache.save(@default_data)
    expect(succeed).to eq(true)
    expect(error).to be_nil

    succeed, data, error = cache.load()
    expect(succeed).to eq(true)
    expect(data).to eq(@default_data)
    expect(error).to be_nil
  end

  it "shouldn't cache process without permissions" do
    cache = CacheSaver.new("/root/fds")

    succeed, error = cache.save(@default_data)
    expect(succeed).to eq(false)
    expect(error).to be_a_kind_of(String)
  end

  it "shoudln't load non existing cache" do
    cache = CacheSaver.new(@default_cache_filepath + "_nonexistent")

    succeed, data, error = cache.load()
    expect(succeed).to eq(false)
    expect(data).to be_nil
    expect(error).to eq("Cache file doesn't exists")
  end

  it "shoudln't load a bad cache format" do
    cache = CacheSaver.new("/etc/hosts")

    succeed, data, error = cache.load()
    expect(succeed).to eq(false)
    expect(data).to be_nil
    expect(error).to be_a_kind_of(String)
  end

end

lib/logstash/filters/ldap.rb
# encoding: utf-8

require "logstash/filters/base"
require "logstash/namespace"

require 'digest/md5'

require 'net/ldap'
require 'rufus-scheduler'

require_relative "buffer/cache_saver"
require_relative "buffer/memory_cache"

class LogStash::Filters::Ldap < LogStash::Filters::Base

  config_name "ldap"

  # Definition of the filter config parameters

  config :identifier_value, :validate => :string, :required => true
  config :identifier_key, :validate => :string, :required => false, :default => "uid"
  config :identifier_type, :validate => :string, :required => false, :default => "posixAccount"

  config :target, :validate => :string, :required => false, :default => "ldap"
  config :attributes, :validate => :array, :required => false, :default => []

  config :host, :validate => :string, :required => true
  config :ldap_port, :validate => :number, :required => false, :default => 389
  config :ldaps_port, :validate => :number, :required => false, :default => 636
  config :use_ssl, :validate => :boolean, :required => false, :default => false

  config :username, :validate => :string, :required => false
  config :password, :validate => :password, :required => false

  config :search_dn, :validate => :string, :required => true

  config :use_cache, :validate => :boolean, :required => false, :default => true
  config :cache_type, :validate => :string, :required => false, :default => "memory"
  config :cache_memory_duration, :validate => :number, :required => false, :default => 300
  config :cache_memory_size, :validate => :number, :required => false, :default => 20000

  config :disk_cache_filepath, :validate => :string, :required => false, :default => nil
  config :disk_cache_schedule, :validate => :string, :required => false, :default => "10m"

  config :enable_error_logging, :validate => :boolean, :required => false, :default => false
  config :no_tag_on_failure, :validate => :boolean, :required => false, :default => false


  # Equivalent to 'initialize' method

  public
  def register

    # Setup some flags

    @SUCCESS = "LDAP_OK"
    @FAIL_PROCESSING = "LDAP_ERROR"
    @NOT_FOUND = "LDAP_NOT_FOUND"

    # Setup some variables

    @attributes_uppercase = @attributes.map(&:upcase)

    @search_all_attributes = (@attributes.length == 0)

    # Check if cache type selected is valid

    if @use_cache
      if @cache_type == "memory"
        @logger.info("Memory cache was selected")
        @Buffer = MemoryCache.new(@cache_memory_duration, @cache_memory_size)
      else
        @logger.warn("Unknown cache type: #{@cache_type}")
        @logger.warn("Cache utilisation will be disable")
        @use_cache = false
      end
    end

    # Set-up of persitant cache

    if @use_cache and !disk_cache_filepath.nil?
      @logger.info("Cache persistance on disk is enabled")
      @disk_cache = CacheSaver.new(@disk_cache_filepath)

      # We load data if any
      succeed, data, error = @disk_cache.load()

      if succeed
        @Buffer.from_obj(data)
        @logger.info("Successfully loaded cache")
      else
        @logger.warn("Failed to load cache : #{error}")
      end

      # We start scheduling save of the cache
      begin
        scheduler.every @disk_cache_schedule do
          data = @Buffer.to_obj()
          succeed, error = @disk_cache.save(data)
          if succeed
            @logger.info("Successfully saved cache")
          else
            @logger.info("Failed to persist cache on disk : #{error}")
          end
        end
      rescue => e
        @logger.warn("Failed to persist cache on disk: #{e.message}")
      end
    end

    # Setup ldap connection

    if @use_ssl
      @ldap = Net::LDAP.new :host => @host,
      :port => @ldaps_port,
      :auth => {
        :method => :simple,
        :username => @username,
        :password => @password.value
      },
      :encryption => {
        :method =>   :simple_tls,
        :tls_options => { verify_mode: OpenSSL::SSL::VERIFY_NONE }
      }
    else
      @ldap = Net::LDAP.new :host => @host,
      :port => @ldap_port,
      :auth => {
        :method => :simple,
        :username => @username,
        :password => @password.value
      }
    end

    # Check the state of the connection

    begin
      if !@ldap.bind()
        raise(@ldap.get_operation_result.error_message)
      end
    rescue Exception => err
      @logger.error("Error while setting-up connection with LDAP server '#{@host}': #{err.message}")
    end
  end

  # This function is called each time an event should be processed

  public
  def filter(event)

    # Get the identifier
    identifier_value = event.sprintf(@identifier_value)

    # Check if the cache is required
    cached = false

    if @use_cache

      # Check if the identifier is cached already via it's hash value
      identifier_hash = hashIdentifier(@host, @port, @identifier_key, identifier_value)

      # Get the cache result
      res = @Buffer.get(identifier_hash)

      # If nothing is in the cache
      if res.nil?

        # Create the LDAP connection and launch the request
        @logger.debug? && @logger.debug("Search for LDAP '#{identifier_value}' element")

        res, exitstatus = ldapsearch(identifier_value)

        # Store the result for futher use
        @Buffer.cache(identifier_hash, res)
      end

    else

      # Create the LDAP connection and launch the request
      @logger.debug? && @logger.debug("Search for LDAP '#{identifier_value}' element")

      res, exitstatus = ldapsearch(identifier_value)
    end

    # Add the result fetched from the database into current event

    res.each{|key, value|
      targetArray = event.get(@target)
      if targetArray.nil?
        targetArray = {}
      end
      targetArray[key] = value
      event.set(@target, targetArray)
    }

    # If there is a problem, set the tag associated if needed

    if !no_tag_on_failure && !exitstatus.nil? && exitstatus != @SUCCESS
      if event.get("tags")
        event.set("tags", event.get("tags") << exitstatus)
      else
        event.set("tags", [exitstatus])
      end
    end

    filter_matched(event)
  end

  # Create an unique hash for an event

  private
  def hashIdentifier(host, port, identifier_key, identifier_value)
    md5 = Digest::MD5.new
    md5.update(host)
    md5.update(port.to_s)
    md5.update(identifier_key)
    md5.update(identifier_value)
    return md5.hexdigest
  end

  # Search LDAP attributes of the object

  private
  def ldapsearch(identifier_value)

    exitstatus = @SUCCESS
    ret = {}

    # Create the request parameters

    object_type_filter = Net::LDAP::Filter.eq("objectclass", @identifier_type)
    identifier_filter = Net::LDAP::Filter.eq(@identifier_key, "#{identifier_value}")

    full_filter = Net::LDAP::Filter.join(identifier_filter, object_type_filter)
    treebase = @search_dn

    # Launch the request

    suceed = false

    begin

      @ldap.search( :base => treebase, :filter => full_filter, :attributes => @attributes) do |entry|
        entry.each do |attribute, values|
          suceed = true
          if @attributes_uppercase.include?(attribute.upcase.to_s) or @search_all_attributes
            ret[attribute] = values.join(" ")
          end
        end
      end

      if !@ldap.get_operation_result.error_message.empty?
        raise(@ldap.get_operation_result.error_message)
      end

    rescue Exception => err
      @logger.error("Error while searching informations: #{err.message}")
      @enable_error_logging && ret["error"] = err.message
      exitstatus  = @FAIL_PROCESSING
      return ret, exitstatus
    end

    # If not, it's probably because the object don't exist

    if !suceed
      @logger.debug? && @logger.debug("Unable to find informations for element '#{identifier_value}'")
      exitstatus = @NOT_FOUND
      return ret, exitstatus
    end

    return ret, exitstatus
  end

end
